File: packages/shared/session-chat-controller/notice-state.ts
import { storageScope } from '@/packages/client-storage';
import type { SessionChatTerminalNotice } from '../session-chat';

const clientStorage = storageScope(['notices']);

export function sessionChatTerminalNoticeDismissKey(notice: SessionChatTerminalNotice | null): string | null {
  return notice ? `${notice.kind}:${notice.detectedAt}` : null;
}

/** What a notice says, independent of when it was detected. */
export function sessionChatTerminalNoticeIdentity(notice: SessionChatTerminalNotice): string {
  return `${notice.kind}:${notice.title}`;
}

/**
 * How long the same screen-state notice stays hidden after being dismissed,
 * even when it comes back under a new `detectedAt`. A usage limit or an expired
 * login lasts far longer than this, and a user who closed the card knows about
 * it; after this long a re-detection is worth mentioning again.
 */
export const NOTICE_REDISPLAY_COOLDOWN_MS = 30 * 60 * 1000;

export interface DismissedNotice {
  /** Exact detection dismissed: `kind:detectedAt`. */
  key: string;
  /** `kind:title` of the dismissed notice. */
  identity: string;
  /** Wall-clock millis of the dismissal. */
  dismissedAt: number;
  /** Whether the cooldown applies: only screen-state notices re-detect continuously. */
  fromScreen: boolean;
}

/**
 * True when `notice` is the detection the user dismissed, or the same
 * screen-state words re-detected within the cooldown.
 */
export function isNoticeDismissed(notice: SessionChatTerminalNotice, dismissed: DismissedNotice | null): boolean {
  if (!dismissed) {
    return false;
  }
  if (sessionChatTerminalNoticeDismissKey(notice) === dismissed.key) {
    return true;
  }
  if (!dismissed.fromScreen || notice.source !== 'screen') {
    return false;
  }
  if (sessionChatTerminalNoticeIdentity(notice) !== dismissed.identity) {
    return false;
  }
  const detectedAt = Date.parse(notice.detectedAt);
  return !Number.isFinite(detectedAt) || detectedAt < dismissed.dismissedAt + NOTICE_REDISPLAY_COOLDOWN_MS;
}

// Per-session dismissed notice (session-chat-verbose-override.ts is the
// pattern). Survives the card unmounting when the host switches surfaces.
const DISMISS_STORAGE_PREFIX = 'ghostex.sessionChat.noticeDismissed.';

export function readStoredDismissedNotice(sessionKey: string | undefined): DismissedNotice | null {
  if (!sessionKey) {
    return null;
  }
  try {
    const raw = clientStorage.getItem(`${DISMISS_STORAGE_PREFIX}${sessionKey}`);
    if (!raw) {
      return null;
    }
    if (!raw.startsWith('{')) {
      // Pre-2026-09-03 entries stored the bare key; they still hide that one detection.
      return { dismissedAt: 0, fromScreen: false, identity: '', key: raw };
    }
    const parsed = JSON.parse(raw) as Partial<DismissedNotice>;
    if (typeof parsed.key !== 'string' || typeof parsed.identity !== 'string') {
      return null;
    }
    return {
      dismissedAt: typeof parsed.dismissedAt === 'number' ? parsed.dismissedAt : 0,
      fromScreen: parsed.fromScreen === true,
      identity: parsed.identity,
      key: parsed.key,
    };
  } catch {
    // Storage disabled by the embedder: dismissal still works, just per-mount.
    return null;
  }
}

export function writeStoredDismissedNotice(sessionKey: string | undefined, dismissed: DismissedNotice): void {
  if (!sessionKey) {
    return;
  }
  try {
    clientStorage.setItem(`${DISMISS_STORAGE_PREFIX}${sessionKey}`, JSON.stringify(dismissed));
  } catch {
    // Quota/private-mode failures must not break the dismiss button.
  }
}

export function dismissedNoticeState(notice: SessionChatTerminalNotice): DismissedNotice {
  return {
    dismissedAt: Date.now(),
    fromScreen: notice.source === 'screen',
    identity: sessionChatTerminalNoticeIdentity(notice),
    key: sessionChatTerminalNoticeDismissKey(notice)!,
  };
}

File: packages/gx-chat-core/src/questions/notice_state.rs
//! Which notice the user closed, and how long closing it keeps the same words off the screen.
//!
//! Port of `packages/shared/session-chat-controller/notice-state.ts`. The stored record is
//! unchanged, including the pre-2026-09-03 form that held only the dismissed key.

use serde::{Deserialize, Serialize};

use crate::questions::model::TerminalNotice;

/// How long the same screen-state notice stays hidden after being dismissed, even when it comes
/// back under a new `detectedAt`.
///
/// A usage limit or an expired login lasts far longer than this, and a user who closed the card
/// knows about it; after this long a re-detection is worth mentioning again.
pub const NOTICE_REDISPLAY_COOLDOWN_MS: i64 = 30 * 60 * 1000;

/// The dismissal one session remembers.
///
/// The field order is the stored key order, because this is written straight to client storage.
#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DismissedNotice {
    /// Wall-clock millis of the dismissal.
    #[serde(default)]
    pub dismissed_at: i64,
    /// Whether the cooldown applies: only screen-state notices re-detect continuously.
    #[serde(default)]
    pub from_screen: bool,
    /// `kind:title` of the dismissed notice.
    pub identity: String,
    /// Exact detection dismissed: `kind:detectedAt`.
    pub key: String,
}

/// The exact detection a dismissal names.
pub fn notice_dismiss_key(notice: Option<&TerminalNotice>) -> Option<String> {
    notice.map(|notice| format!("{}:{}", notice.kind, notice.detected_at))
}

/// What a notice says, independent of when it was detected.
pub fn notice_identity(notice: &TerminalNotice) -> String {
    format!("{}:{}", notice.kind, notice.title)
}

/// True when `notice` is the detection the user dismissed, or the same screen-state words
/// re-detected within the cooldown.
pub fn is_notice_dismissed(notice: &TerminalNotice, dismissed: Option<&DismissedNotice>) -> bool {
    let Some(dismissed) = dismissed else {
        return false;
    };
    if notice_dismiss_key(Some(notice)).as_deref() == Some(dismissed.key.as_str()) {
        return true;
    }
    if !dismissed.from_screen || notice.source != "screen" {
        return false;
    }
    if notice_identity(notice) != dismissed.identity {
        return false;
    }
    match parse_iso_millis(&notice.detected_at) {
        // An unreadable stamp cannot be compared, so the cooldown holds rather than flashing the
        // card the user just closed.
        None => true,
        Some(detected_at) => detected_at < dismissed.dismissed_at + NOTICE_REDISPLAY_COOLDOWN_MS,
    }
}

/// The record written when the user closes a notice.
pub fn dismissed_notice_state(notice: &TerminalNotice, now_ms: i64) -> DismissedNotice {
    DismissedNotice {
        dismissed_at: now_ms,
        from_screen: notice.source == "screen",
        identity: notice_identity(notice),
        key: notice_dismiss_key(Some(notice)).unwrap_or_default(),
    }
}

/// Reads a stored dismissal, tolerating the record shape that predates the cooldown.
pub fn decode_dismissed_notice(raw: &str) -> Option<DismissedNotice> {
    if raw.is_empty() {
        return None;
    }
    if !raw.starts_with('{') {
        // Pre-2026-09-03 entries stored the bare key; they still hide that one detection.
        return Some(DismissedNotice {
            dismissed_at: 0,
            from_screen: false,
            identity: String::new(),
            key: raw.to_string(),
        });
    }
    serde_json::from_str(raw).ok()
}

/// `Date.parse` for the ISO-8601 stamps gxserver writes, in epoch milliseconds.
///
/// One rule for the whole crate since 2026-09-22 (`crate::jstime`). The core reads no timezone
/// here, so a stamp without an offset is still taken as UTC; gxserver always writes `Z`.
fn parse_iso_millis(text: &str) -> Option<i64> {
    crate::jstime::parse_iso_millis_utc(text).map(|millis| millis as i64)
}
